### PR TITLE
Implement algorithm and tests for word searching in a grid of letters

### PR DESCRIPTION
**Pull request by veronikakolejak**

### CODE REVIEW

File: Assignments/Assignment3/WordSearch.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package Assignment3;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+
+public class WordSearch {
+
+  private static int[] deltaX = {-1, -1, 0, 1, 1, 1, 0, -1};
+  private static int[] deltaY = {0, 1, 1, 1, 0, -1, -1, -1};
```
Comment (karineek) [inline Assignments/Assignment3/WordSearch.java:10]: Better to use a loop without of using extra memory. Since we anyhow run from -1 to 1 per x and y, just set int x = -1 (instead of 0) and iterate till x < 2.

File: Assignments/Assignment3/WordSearchTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        }
+      }
+      return false;
+    }
+  };
+
+  public void testFindWords() throws Exception {
+    /* Grid example used for this test */
+    // T C D
+    // A A R
```
Comment (karineek) [inline Assignments/Assignment3/WordSearchTest.java:38]: Shall add the opposite test, where there are more rows the cols.

File: Assignments/Assignment3/WordSearchTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class WordSearchTest extends TestCase {
+  public static Dictionary dict = new Dictionary() {
+
+    List<String> words = new ArrayList<>(Arrays.asList(
+      "car", "card", "cart", "cat", "cardc", "tata", "a", "ar", "pines", "dates", "pint"
```
Comment (karineek) [inline Assignments/Assignment3/WordSearchTest.java:16]: Shall also add these tests:
1. a word that contains all the letters, 2. a word that starts to the end of the grid (d,r and o,v for example), and 3. a word that contains duplications that cannot be, ass ACAC.

File: Assignments/Assignment3/WordSearch.java (hunk 1 above)
Comment (simon-frankau) [reply Assignments/Assignment3/WordSearch.java:10]: I agree. The memory usage isn't important here, but a couple of small for loops is easier to understand than having to read this array and make sure you haven't typo'd one of the cases.

File: Assignments/Assignment3/WordSearch.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      words.add(currentWord);
+    }
+
+    for (int i = 0; i < deltaX.length; i++) {
+      int newX = curX + deltaX[i];
+      int newY = curY + deltaY[i];
+
+      if (newX >= 0 && newX < visited[0].length && newY >= 0 && newY < visited.length && !visited[newY][newX]) {
+        findWordsHelper(grid, visited, newX, newY, dictionary, currentWord, words);
+        visited[newY][newX] = false;
```
Comment (simon-frankau) [inline Assignments/Assignment3/WordSearch.java:50]: It's a simpler, cleaner invariant if you make sure that findWordsHelper leaves the "visited" array as it found it. So, rather than clean it here, clean it inside the called method. If you do this, you won't need to create a new "visited" array in every iteration of the findWords loop.

File: Assignments/Assignment3/WordSearchTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    assertThat(words, not(hasItems("")));
+
+    /* Check that it does not use letters not present in the grid */
+    assertThat(words, not(hasItems("hey")));
+
+    /* Check that it does not output two appended dictionary words */
+    assertThat(words, not(hasItems("aar")));
+
+    /* Check that it does not use letters not present in the dictionary */
+    assertThat(words, not(hasItems("tar")));
```
Comment (simon-frankau) [inline Assignments/Assignment3/WordSearchTest.java:62]: Rather than checking individually if it contains a specific word or not, why not just check that the found words equals precisely the set of words you expect?

File: Assignments/Assignment3/WordSearchTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    List<String> words = WordSearch.findWords(4, 4, grid, dict);
+
+    /* Check for correct valid words from dictionary */
+    assertThat(words, hasItems("pines"));
+
+    /* Check that it does not find words whose letters are not adjacent or sequentially adjacent */
+    assertThat(words, not(hasItems("dates", "pint")));
+  }
+}
```
Comment (simon-frankau) [inline Assignments/Assignment3/WordSearchTest.java:81]: It looks like the test dictionaries have some good words for exercising corner-case behaviour, and between that and the suggestions by karineek you should have pretty good coverage.

**Review by simon-frankau — APPROVED** [review]
I suggest a couple of minor tweaks, but it looks like it does the job. What algorithmic complexity do you think the algorithm is?
